Consider confidence values in 'ensure_consistent_bilou_tagging'

tabergma opened the pull request:
**Proposed changes**:
The method `ensure_consistent_bilou_tagging` now also considers the confidence values of the predicted tags
when updating the BIlOU tags.

closes https://github.com/RasaHQ/rasa/issues/6039

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

tabergma commented on rasa/nlu/utils/bilou_utils.py:
```diff
@@ -271,6 +277,32 @@ def ensure_consistent_bilou_tagging(predicted_tags: List[Text]) -> List[Text]:
+                )
+
+                # the entity should have the tag with the highest average confidence
+                tag = max(avg_confidence_per_tag, key=avg_confidence_per_tag.get)
```
@Ghostvv do you think this is enough or should we also take into account how many tags within the entity are predicted as e.g. person or location?

Ghostvv replied:
I think how many tags is more robust measure. E.g. person 2 times with 0.89 and location 1 time with 0.95. Maybe it is possible to come up with some kind of F1 score between these values

Ghostvv replied:
also, it'd be great to log this mistakes somehow, so that users can be pointed out where more data is required

tabergma replied:
👍

tabergma replied:
updated the metric

Ghostvv commented on rasa/nlu/utils/bilou_utils.py:
```diff
@@ -283,10 +311,113 @@ def ensure_consistent_bilou_tagging(predicted_tags: List[Text]) -> List[Text]:
+    tag = max(score_per_tag, key=score_per_tag.get)
+    avg_confidence = avg_confidence_per_tag[tag]
+
+    return tag, avg_confidence
```
should the score be used as "confidence" then?

tabergma replied:
might make sense, the score would be for sure lower compared to the avg confidence score

Ghostvv replied:
but that's good I guess, since it is kind of artificial substitution

Ghostvv reviewed (commented):
not sure how troublesome to implement it, but is it possible to output original tags during `rasa test` as well?

tabergma commented:
You mean the tags before they got updated? mmhh... not that easy, as the tags are converted into proper entities and I am not sure how you would add the "old" tags in there.

Ghostvv commented:
Let's create an issue for it to discuss it. So that this PR is not blocked

tabergma commented:
https://github.com/RasaHQ/rasa/issues/6647

Ghostvv commented:
Could you please add the description of this substitution process to docs, please?

tabergma commented:
Where should it go? We do not have a section for it as far as i know.

Ghostvv commented:
hm, it seems that we don't have a place where we describe bilou tagging scheme. I think we should add it some where

tabergma commented:
@erohmensing tagging you as a review as I added a section about BILOU tagging to the docs. As the information about `NLU Training Data` was already included in `Training Data Format` I removed it and kept just the entity related stuff. Let me know if that does not work and the section should be moved somewhere else.

Ghostvv reviewed (commented):
the part about BILOU tags looks good.
I wouldn't remove `nlu-training-data.mdx` as part of this PR. To avoid confusion, let's do it in a separate PR

tabergma commented:
followed the suggestion of @Ghostvv and added the page about nlu training data again and just added the bilou section to that page

erohmensing commented on changelog/6039.improvment.md:
```diff
@@ -0,0 +1,2 @@
+The method `ensure_consistent_bilou_tagging` now also considers the confidence values of the predicted tags
+when updating the BIlOU tags.
```
```suggestion
when updating the BILOU tags.
```

erohmensing commented on docs/docs/nlu-training-data.mdx:
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
 [Forms](forms.mdx) or use [Custom Actions](custom-actions.mdx) to extract the corresponding entity directly from the tracker.
+
+
+### Entity with BILOU tagging
```
Placement is fine for now, i'll find a spot with it when reviewing your updates to this page :) 
```suggestion
### BILOU Entity Tagging
```

erohmensing commented on docs/docs/nlu-training-data.mdx:
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
+
+### Entity with BILOU tagging
+
+[DIETClassifier](./components.mdx#dietclassifier) and [CRFEntityExtractor](./components.mdx#crfentityextractor)
+have the option `BILOU_flag`.
+`BILOU` is short for `Beginning, Inside, Last, Outside, and Unit-length` and refers to a tagging schema that can be
+used by the machine learning model.
```
```suggestion
The [DIETClassifier](./components.mdx#dietclassifier) and [CRFEntityExtractor](./components.mdx#crfentityextractor)
have the option `BILOU_flag`, which refers to a tagging schema that can be
used by the machine learning model when processing entities.
`BILOU` is short for Beginning, Inside, Last, Outside, and Unit-length.

```

erohmensing commented on docs/docs/nlu-training-data.mdx:
````diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
+[Alex]{"entity": "person"} is going with [Marty A. Rick]{"entity": "person"} to [Los Angeles]{"entity": "location"}.
+```
+
+is first split into a list of tokens and then the machine learning model applies the tagging schema
+depending on the option `BILOU_flag` during training.
````
```suggestion
is first split into a list of tokens. Then the machine learning model applies the tagging schema
as shown below, depending on the value the `BILOU_flag`:
```

erohmensing commented on docs/docs/nlu-training-data.mdx:
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
+is first split into a list of tokens and then the machine learning model applies the tagging schema
+depending on the option `BILOU_flag` during training.
+
+| token   | BILOU tagging schema | normal tagging schema |
```
```suggestion
| token   | `BILOU_flag = true`  | `BILOU_flag = false`  |
```

erohmensing commented on docs/docs/nlu-training-data.mdx:
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
+performance of the machine learning model when predicting entities.
+
+:::note
+When the option `BILOU_flag` is set to `True`, it might happen that the model is predicting inconsistent BILOU tags, e.g.
```
```suggestion
When the option `BILOU_flag` is set to `True`, the model may predict inconsistent BILOU tags, e.g.
```

erohmensing commented on docs/docs/nlu-training-data.mdx:
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
+The BILOU tagging schema is richer compared to the normal tagging schema. It may help to improve the
+performance of the machine learning model when predicting entities.
+
+:::note
```
```suggestion
:::note inconsistent bilou tags
```

erohmensing reviewed (commented):
👍

tabergma commented:
Applied all review comments, thanks for the comments!

@erohmensing or @Ghostvv can someone please approve so that we can merge this PR? Thanks.

tabergma commented:
oh no, just realised that we will not be able to merge as a lot of tests will all fail due to the convert issue
